doc/source/development/roadmap-indexing-views.rst

```diff
@@ -0,0 +1,310 @@
+   >>> df2 = df[["A"]]
+   >>> df2.iloc[:, 0] = 10
+
+Did the user intend to modify ``df`` when they modified ``df2`` (setting aside
+issues with the current implementation)? In other words, if we had a perfectly
+consistent world where indexing the columns always returned views or always
+returned a copy, does the code above imply that the user wants to mutate ``df``?
+
+There are two possible behaviours the user might intend:
+
+1. I know my subset might be a view of the original and I want to modify the
+   original as well.
+2. I just want to modify the subset without modifying the original.
+
+Today, pandas’ inconsistency means neither of these workflows is really
+possible. The first is difficult, because indexing operations often (though not
+always) return copies, and even when a view is returned you sometimes get a
+``SettingWithCopyWarning`` when mutating. The second is somewhat possible, but
+requires many defensive copies (to avoid ``SettingWithCopyWarning``, or to
+ensure that you have a copy when a view was returned).
+
+Proposal Summary
+----------------
+
+For these reasons (consistency, performance, code clarity), we propose three
+changes:
+
+1. Indexing always returns a view when possible. This means that indexing
+   columns of a dataframe always returns a view
+   (https://github.com/pandas-dev/pandas/pull/33597), and indexing rows may
+   return a view, depending on the type of the row indexer.
+2. We implement Error-on-Write (explained below)
+3. We provide APIs for explicitly marking a DataFrame as a “mutable view”
+   (mutating the dataframe would mutate its parents) and copying a dataframe
+   only if needed to avoid concerns with mutating other dataframes (i.e. it is
+   not a view on another dataframe).
+
+The intent is to capture the performance benefits of views, while allowing users
+to explicitly choose the behavior they want for inplace operations that might
+mutate other dataframes. This essentially makes returning views an internal
+optimization, without the user needing to know if the specific indexing
+operation would return a view or a copy.
+
+Taking the example from above, if the user wants to make use of the fact that
+``df2`` is a view to modify the original ``df``, they would write:
+
+.. code-block:: python
+
+   # Case 1: user wants mutations of df2 to be reflected in df
+   >>> df = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+   >>> df2 = df[["A"]].as_mutable_view() # name TBD
+   >>> df2.iloc[:, 0] = 10 >>> df.iloc[0, 0] # df was mutated 10
+
+For the user who wishes to not mutate the parent, we require that the user
+explicitly break the reference from ``df2`` to ``df`` by implementing “Error on Write”.
+
+.. code-block:: python
+
+   # Case 2: The user does not want mutating df2 to mutate df, via EoW
+   >>> df = pd.DataFrame({"A": [1, 2], "B": [3, 4]})
+   >>> df2 = df[["A"]]
+   >>> df2.iloc[0, 0] = 10
+   ValueError("error on write to subset of other dataframe")
```

Comment:
Maybe we should create a new exception type instead of `ValueError` (with name TBD) ?

Reply:
Sure. Will call it `MutableViewError`, but happy for better names.

Reply:
Kind of interesting that numpy uses a `ValueError` if something isn't writeable:
```
[ins] In [1]: import numpy as np

[ins] In [2]: arr = np.array([1, 2, 3])

[ins] In [3]: arr.flags.writeable = False

[ins] In [4]: arr[0] = 9
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-1804f9f074cb> in <module>
----> 1 arr[0] = 9

ValueError: assignment destination is read-only
```
Possibly even just `ReadOnlyError`?